addon/components/frost-autocomplete.js

```diff
@@ -0,0 +1,316 @@
+      this._setFocusedIndex(upArrow)
+    }
+  },
+
+  _setFocusedIndex (upArrow) {
+    const items = this.get('items')
+    if (isEmpty(items)) {
+      return
+    }
+
+    const selectedItemIndex = this.get('focusedIndex') || 0
+
+    const newSelectedIndex = (
+      upArrow ? Math.max(0, selectedItemIndex - 1) : Math.min(items.length - 1, selectedItemIndex + 1)
+    )
+
+    if (newSelectedIndex !== selectedItemIndex) {
+      this.set('focusedIndex', newSelectedIndex)
+    }
+  },
+
+  @readOnly
+  @computed('width')
+
+  /**
+   * Get a style string based on the presence of some properties
+   * @param {String} width the width property specified
+   * @returns {String} the completed style string
+   */
+  style (width) {
+    let styles = ''
+
+    // if a property is not falsy, append it to the style string
+    // note that in the width case, we want the component interface to have absolute power over the width
+    // so it will override any max or win widths to ensure ultimate control
+    if (width) styles += `width: ${width}px; max-width: initial; min-width:initial; `
+    return Ember.String.htmlSafe(styles)
+  },
+
+  @readOnly
+  @computed('disabled', 'tabIndex')
+  /**
+   * Get appropriate tab index
+   * disabled state changes.
+   * @param {Boolean} disabled - whether or not input is disabled
+   * @param {Number} tabIndex - tab index
+   * @returns {Number} tab index
+   */
+  computedTabIndex (disabled, tabIndex) {
+    return disabled ? -1 : tabIndex
+  },
+
+  @readOnly
+  @computed('opened', '_userInput', 'filter', 'focused')
+  _opened (opened, userInput, filter, focused) {
+    return (opened || focused) && userInput && !isEmpty(filter)
+  },
+
+  @readOnly
+  @computed('data', 'filter', 'onInput')
+  items (data, filter, onInput) {
+    if (isEmpty(data)) {
+      return []
+    }
+
+    filter = filter ? filter.toLowerCase() : null
+
+    return data.filter((item) => {
+      if (isEmpty(filter)) {
+        return true
+      }
+
+      const label = item.label || ''
+
+      if (this.isFilteredItem(label.toLowerCase(), filter)) {
+        return true
+      }
+    })
+  },
+
+  init () {
+    this._super(...arguments)
+
+    if (!isEmpty(this.selectedValue)) {
+      this.set('internalSelectedItem', {value: this.selectedValue})
+    }
+  },
+
+  didInsertElement () {
+    this._super(...arguments)
+
+    // We need jQuery instance of components root DOM node to hand off to
+    // dropdown so it can position itself properly relative to the select
+    this.set('$element', this.$())
+  },
+
+  // == DOM Events ============================================================
+
+  _onClick: on('click', function (e) {
+    if (!this.get('disabled')) {
+      this.toggleProperty('opened')
+    }
+    if (this.onClick) {
+      this.onClick(e)
+    }
+  }),
+
+  _onKeyDown: on('keyDown', function (e) {
+    if (
+      [DOWN_ARROW, UP_ARROW].indexOf(e.keyCode) !== -1 &&
+      !this.get('openend')
```

Comment:
This appears to be spelled incorrectly.   If indeed it is why did no tests catch this?  Write some additional tests to cover this condition as I would not have expected them to pass since referencing an incorrect variable name.

Reply:
I actually don't need the opened check, the typo actually let it work! good times :D. Thanks

Reply:
When you removed the `&& !this.get('openend')` code did you modify a test case as well?  If not then your test coverage is/was still not accurate as either the inclusion or the removal of this code should have caused a test to fail in one of the conditions or the other.

Reply:
When i tried it with it spelled correctly, everything worked fine, except the cursor in the input would move left or right depending on the arrow. I could write a test around this to make sure it's not moving the cursor

Reply:
The code was initially spelled incorrectly and things seemed to be working correctly via what I'm guessing was visual inspection only, since no tests were failing (or the tests were also incorrect), correct?

Then you corrected the spelling and things containing to work correctly via what I'm guessing was visual inspection only, since not tests were failing (or the tests were also incorrect), correct?

Since no change in behavior was observed you then removed the `&& !this.get('openend') logic all together then?

When it was spelled incorrectly did you notice this same behavior:

> everything worked fine, except the cursor in the input would move left or right depending on the arrow.

or did this behavior only begin occurring once you spelled it correctly?

And when you say

> except the cursor in the input would move left or right depending on the arrow

are you referring to the behavior of the component when being used or within test cases?


I'd like to take a step back and getting a broader understanding of what is going on.  I would have expected that the mis-spelling to cause a test to fail. Since it did not it either means that there is no test coverage for this condition or that the test is also wrong.  Likewise, when you remove this logic, I would expect tests to fail and since they did not there is either no test coverage for this condition or the test is wrong.  Where I'm really confused though is the introduction of the conversation around the arrow keys and where they fit it in (see my previous question above).  At first glance I would expect the pressing of an arrow key to cause the cursor to move....

Reply:
the typo'd part always resulted in true. This is why it was working correctly. I just removed the dead logic. It will make sure the dropdown is open using up/down in my case as long as the input has already had actual interacted with the input